README.txt
ABOUT

Jailkit is a set of utilities to limit user accounts to specific files 
using chroot() and or specific commands. Setting up a jail is a lot 
easier using these utilities.

See 'man jailkit' for more detailed info on the different utilities 
in this package. If you have not yet installed jailkit you can see 
this man page using 'man man/jailkit.8'.

REQUIREMENTS

The most elementary tools are written in C and only need libc and 
libpthreads (available on almost any Linux system and most other posix
compliant systems).

The scripts are written in python, so they need python installed.

COPYRIGHT

Jailkit is an open source project written by Olivier Sessink. It is 
released under a modified BSD licence.

EXAMPLE

Suppose you need to create an account 'test' that can do sftp and scp
only. You want it in a jail called /home/sftproot where it will have 
a homedirectory /home/test

#initialise the jail
mkdir /home/sftproot
jk_init /home/sftproot jk_lsh
jk_init /home/sftproot sftp
jk_init /home/sftproot scp
# create the account
adduser test
jk_jailuser /home/sftproot test
# edit the jk_lsh configfile in the jail (man jk_lsh)
# you can use every editor you want, I chose 'joe'
joe /home/sftproot/etc/jailkit/jk_lsh.ini
# now restart jk_socketd
killall jk_socketd
jk_socketd
# test the account
sftp test@localhost
# check the logs if everything is correct
tail /var/log/daemon.log /var/log/auth.log

NOTES ON VARIOUS APPLICATIONS

cvs - cvs needs /tmp/ for temporary files
procmail - needs /dev/null

COPYRIGHT

Copyright (C) 2003, 2004, 2005, Olivier Sessink

Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.


debian/changelog
jailkit (2.15-1) lucid; urgency=low

  * use -pthreads if available instead of -lpthread
  * allow symlinks for /bin /lib and /sbin if they are moved into /usr/ (as fedora 17 has)

 -- Olivier Sessink <olivier@cort.fakenet>  Thu, 07 Jun 2012 23:20:26 +0200

jailkit (2.14-1) lucid; urgency=low

  * move jk_check to hashlib
  * fix for the prefix directory of jk_check
  * enable jk_chrootsh to be called su

 -- Olivier Sessink <olivier@cort.fakenet>  Thu, 28 Apr 2011 10:25:24 +0200

jailkit (2.13-1) lucid; urgency=low

  * fixes a regression in 2.12 - INIPREFIX was updated incorrectly in some cases 

 -- Olivier Sessink <olivier@cort.fakenet>  Sun, 10 Oct 2010 22:35:24 +0200

jailkit (2.12-1) lucid; urgency=low

  * jk_init and jk_cp can now resolve binaries using PATH

 -- Olivier Sessink <olivier@cort.fakenet>  Sun, 12 Sep 2010 21:30:23 +0200

jailkit (2.11-1) hardy; urgency=low

  * mostly documentation updates
  * minor Solaris specific improvements
  * fixed possible jk_lsh crash

 -- Olivier <olivier@cort.fakenet>  Sun, 07 Feb 2010 17:23:16 +0100

jailkit (2.10-1) hardy; urgency=low

  * 2.9 fix was incompatible with jk_jailuser, fixed now
  * compiler warning fixes

 -- Olivier <olivier@cort.fakenet>  Thu, 22 Oct 2009 13:31:18 +0200

jailkit (2.9-1) hardy; urgency=low

  * fix symlink handling, if a symlink inside the jail points to the real system jk_init and jk_cp could write files into the real system

 -- Olivier <olivier@cort.fakenet>  Wed, 14 Oct 2009 23:07:33 +0200

jailkit (2.8-1) hardy; urgency=low

  * support for capabilities to replace the setuid bit
  * minor Solaris compatibility fixes

 -- Olivier <olivier@cort.fakenet>  Thu, 20 Aug 2009 12:51:05 +0200

jailkit (2.7-1) hardy; urgency=low

  * fix a regression in jk_chrootsh and jk_uchroot for interactive sessions

 -- Olivier <olivier@cort.fakenet>  Sun, 05 Apr 2009 20:20:38 +0200

jailkit (2.6-1) hardy; urgency=low

  * some minor code cleanups and improvements
  * some fixes for Solaris compatibility

 -- Olivier <olivier@cort.fakenet>  Wed, 01 Apr 2009 21:13:20 +0200

jailkit (2.5-1) unstable; urgency=low

  * fixed an issue in jk_cp and jk_init that symlinks to directories were sometimes copied as directories instead of symlinks
  * improved some documentation
  * added a -j option to all utilities to improve consistency

 -- Olivier Sessink <olivier@aria.fakenet>  Fri, 07 Dec 2007 13:53:52 +0100

jailkit (2.4-1) unstable; urgency=low

  * added a new jk_uchroot utility
  * fixed a rare crash in configuration file parsing 
  * added new jk_init configuration file options
  * added new options to jk_cp and jk_init to retain ownership when copying files into the jail
  * improved the documentation
  * improved several chroot security checks 
  * fixed some build issues on OpenBSD
  * improved error reporting

 -- Olivier Sessink <olivier@cort.fakenet>  Tue, 17 Jul 2007 20:32:41 +0200

jailkit (2.3-1) unstable; urgency=low

  * various small fixes for the new jk_update utility
  * added a -k option to jk_cp, jk_init and jk_update to use hardlinks instead of copying
  * various fixes to install jailkit in a different location

 -- Olivier Sessink <olivier@aria.fakenet>  Thu, 28 Dec 2006 13:51:00 +0100

jailkit (2.2-1) unstable; urgency=low

  * Jailkit 2.2 released with a new utility jk_update to keep jails up-to-date
    and various improvements on *BSD operating systems

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Fri, 8 Dec 2006 00:00:00 +0200

jailkit (2.1-2) unstable; urgency=low

  * tarballs are uploaded again, there was a Makefile issue

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Fri, 8 Sep 2006 00:00:00

jailkit (2.1-1) unstable; urgency=low

  * improved logging
  * allowed sub-second intervals for jk_socketd
  * improved various checks in jk_jailuser
  * improved handling of recursive directories in jk_cp
  * new utility jk_list to list all jailed processes

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Thu, 7 Sep 2006 00:00:00

jailkit (2.0-1) unstable; urgency=low

  * fixes to allow jk_chrootsh to function on systems that use a name service
    caching daemon, or ldap authentication and others
  * some options for jk_chrootsh to relax some strict checks
  * replaced jk_addjailuser with jk_jailuser, which should work on more
    platforms
  * added the DEFAULT section to jk_lsh.ini and jk_chrootsh.ini to allow
    easier configuration

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Wed, 9 Aug 2006 00:00:00

jailkit (1.3-1) unstable; urgency=low

  * fixes some issues with Gentoo Linux
  * fixes some FreeBSD issues
  * fixes option --force for jk_init
  * better defaults in jk_init.ini
  * much more verbose error reporting by jk_lsh and jk_chrootsh

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Thu,  5 Aug 2004 19:22:41

jailkit (1.2-1) unstable; urgency=low

  * new upstream 1.2 release
  * mainly *BSD and MacOSX build related fixes

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Thu, 26 Aug 2004 21:34:01 +0200

jailkit (1.1-2) unstable; urgency=low

  * adding lintian override file, package now is lintian clean

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Sun, 15 Aug 2004 11:47:44 +0200

jailkit (1.1-1) unstable; urgency=low

  * new upstream 1.1 release
  * fixes some leftovers from /usr/bin/ to /usr/sbin/ transition
  * fixes minor bug in jk_init

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Fri, 13 Aug 2004 21:23:43 +0200

jailkit (1.0-1) unstable; urgency=low

  * new upstream 1.0 release
  * jk_adduser improvements
  * documentation improvements
  * moved all binaries to /usr/sbin/

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Thu,  5 Aug 2004 19:22:41 +0200

jailkit (0.9-1) unstable; urgency=low

  * new upstream 0.9 release
  * jk_cp and jk_init speed improvements

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Sun, 23 May 2004 21:28:17 +0200

jailkit (0.8-1) unstable; urgency=low

  * new upstream 0.8 release
  * adds wildcard expansion (*,?,~, etc.) to jk_lsh

src/jk_uchroot.c
/*
 * the jailkit chroot() shell
 * this program does a safe chroot() and then executes the shell
 * that the user has within that new root (according to newroot/etc/passwd)
 *
 * I tried to merge some of the ideas from chrsh by Aaron D. Gifford, 
 * start-stop-daemon from Marek Michalkiewicz and suexec by the Apache 
 * group in this shell
 *

Copyright (c) 2003, 2004, 2005, 2006, 2007, 2008, 2009 Olivier Sessink
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions 
are met:
  * Redistributions of source code must retain the above copyright 
    notice, this list of conditions and the following disclaimer.
  * Redistributions in binary form must reproduce the above 
    copyright notice, this list of conditions and the following 
    disclaimer in the documentation and/or other materials provided 
    with the distribution.
  * The names of its contributors may not be used to endorse or 
    promote products derived from this software without specific 
    prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
POSSIBILITY OF SUCH DAMAGE.
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <pwd.h>
#include <grp.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <errno.h>
#include <syslog.h>
#include <limits.h>
#include <fcntl.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif
#ifdef HAVE_LIBERTY_H
#include <liberty.h>
#endif
#ifdef HAVE_SYS_CAPABILITY_H
#include <sys/capability.h>
#endif
/*#define DEBUG*/

#ifdef DEBUG
#define DEBUG_MSG printf
#else
#define DEBUG_MSG(args...)
 /**/
#endif

#define PROGRAMNAME "jk_uchroot"
#define CONFIGFILE INIPREFIX"/jk_uchroot.ini"

#include "jk_lib.h"
#include "utils.h"
#include "iniparser.h"
#include "passwdparser.h"

/* doesn't compile on FreeBSD without this */
extern char **environ;

static void print_usage() {
	printf(PACKAGE" "VERSION"\nUsage: "PROGRAMNAME" -j jaildir -x executable -- [executable options]\n");
	printf("\t-j|--jail jaildir\n");
	printf("\t-x|--exec executable\n");
	printf("\t-h|--help\n");
	printf(PROGRAMNAME" logs all errors to syslog, for diagnostics check your logfiles\n");
}

static int have_capabilities(void) {
#ifdef HAVE_CAP_GET_PROC
	cap_t caps = cap_get_proc();
	if (caps) {
		cap_flag_value_t value_p;
		cap_get_flag(caps, CAP_SYS_CHROOT, CAP_EFFECTIVE,&value_p);
		cap_free(caps);
		return (value_p);
	}
#endif  /*HAVE_CAP_GET_PROC*/
	return 0;
}

/* check basics */
/* parse arguments */
/* parse configfile */
/* check user info */
/* check jail */
/* do chroot call */
int main(int argc, char **argv) {
	char *jail = NULL;
	char *user = NULL;
	char *executable = NULL;
	struct passwd *pw=NULL;
	struct group *gr=NULL;
	Tiniparser *parser=NULL;
	char **newargv=NULL;
	char **allowed_jails = NULL;
	unsigned int skip_injail_passwd_check=0;
	unsigned int i;
	char *tmp;
	unsigned int use_capabilities=0;

	openlog(PROGRAMNAME, LOG_PID, LOG_AUTH);
	
	/* check if it us that the user wants */
	{
		char *tmp = strrchr(argv[0], '/');
		if (!tmp) {
			tmp = argv[0];
		} else {
			tmp++;
		}
		if (strcmp(tmp, PROGRAMNAME) && (tmp[0] != '-' || strcmp(&tmp[1], PROGRAMNAME))) {
			DEBUG_MSG("wrong name, tmp=%s, &tmp[1]=%s\n", tmp, &tmp[1]);
			syslog(LOG_ERR, "abort, "PROGRAMNAME" is called as %s", argv[0]);
			exit(1);
		}
	}
	
	/* now test if we are setuid root (the effective user id must be 0, and the real user id > 0 */
#ifndef DEVELOPMENT
	if (geteuid() != 0) {
		if (have_capabilities()) {
			use_capabilities=1;
		} else {
			syslog(LOG_ERR, "abort, effective user ID is not 0, possibly "PROGRAMNAME" is not setuid root");
			exit(11);
		}
	}
#endif
	if (getuid() == 0) {
		syslog(LOG_ERR, "abort, "PROGRAMNAME" is run by root, which does not make sense because user root can use the chroot utility");
		exit(12);
	}
	
	

	DEBUG_MSG("get user info\n");
	/* get user info based on the users name and not on the uid. this enables support
	for systems with multiple users with the same user id*/
	tmp = getenv("USER");
	if (tmp && strlen(tmp)) {
		user = strdup(tmp);
	}
	if (user) {
		pw = getpwnam(user);
	} else {
		pw = getpwuid(getuid());
	}

	if (!pw) {
		syslog(LOG_ERR, "abort, failed to get user information for user ID %d: %s, check /etc/passwd", getuid(), strerror(errno));
		exit(13);
	}
	if (!pw->pw_name || strlen(pw->pw_name)==0) {
		syslog(LOG_ERR, "abort, got an empty username for user ID %d: %s, check /etc/passwd", getuid(), strerror(errno));
		exit(13);
	}
	if (user && strcmp(user,pw->pw_name)!=0) {
		syslog(LOG_ERR, "abort, asked for user %s, got user info for %s", user, pw->pw_name);
		exit(13);
	}
	if (pw->pw_uid != getuid()) {
		syslog(LOG_ERR, "abort, started by user ID %d, got user info %s with user ID %d,", getuid(), pw->pw_name, pw->pw_uid);
		exit(13);
	}
	gr = getgrgid(getgid());
	if (!gr) {
		syslog(LOG_ERR, "abort, failed to get group information for group ID %d: %s, check /etc/group", getgid(), strerror(errno));
		exit(13);
	}


	{
		int c=0;
		while (c != -1) {
			int option_index = 0;
			static struct option long_options[] = {
				{"jail", required_argument, NULL, 'j'},
				{"executable", required_argument, NULL, 'x'},
				{"help", no_argument, NULL, 'h'},
				{"version", no_argument, NULL, 'V'},
				{NULL, 0, NULL, 0}
			};
		 	c = getopt_long(argc, argv, "j:x:hv",long_options, &option_index);
			switch (c) {
			case 'j':
				jail = ending_slash(optarg);
				DEBUG_MSG("argument jail='%s', ending_slash returned '%s'\n",optarg,jail);
				break;
			case 'x':
				executable = strdup(optarg);
				break;
			case 'h':
			case 'V':
				print_usage();
				exit(1);
			}
		}
		/* construct the new argv from all leftover options */
		newargv = malloc0((2 + argc - optind)*sizeof(char *));
		newargv[0] = executable;
		c = 1;
		while (optind < argc) {
			newargv[c] = strdup(argv[optind]);
			c++;
			optind++;
		}
		
		if (jail == NULL) {
			printf("ERROR: No jail path specified. Use -j or --jail\n");
			print_usage();
			exit(1);
		}
	}
	/* make sure the jailkit config directory is owned root:root and not writable for others */
	if ( (testsafepath(INIPREFIX, 0, 0) &~TESTPATH_GROUPW) != 0 ) {
		syslog(LOG_ERR, "abort, jailkit configuration directory "INIPREFIX" is not safe; it should be owned 0:0 and not writable for others");
		exit(14);
	}
	parser = new_iniparser(CONFIGFILE);
	if (parser) {
		/* first check for a section specific for this user, then for a group section, else a DEFAULT section */
		char *groupsec, *section=NULL, buffer[1024]; /* openbsd complains if this is <1024 */
		groupsec = strcat(strcpy(malloc0(strlen(gr->gr_name)+7), "group "), gr->gr_name);
		if (iniparser_has_section(parser, pw->pw_name)) {
			section = strdup(pw->pw_name);
		} else if (iniparser_has_section(parser, groupsec)) {
			section = groupsec;
		} else if (iniparser_has_section(parser, "DEFAULT")) {
			section = strdup("DEFAULT");
		}
		if (section != groupsec) free(groupsec);
		if (section) {
			/* from this section, retrieve the options 
			 - which jails are allowed
			 - which shell to use
			 - if the user has to be in <jail>/etc/passwd (only if shell is given)
			 */
			unsigned int pos = iniparser_get_position(parser) - strlen(section) - 2;
			if (iniparser_get_string_at_position(parser, section, "allowed_jails", pos, buffer, 1024) > 0) {
				DEBUG_MSG("found allowed_jails=%s\n",buffer);
				allowed_jails = explode_string(buffer, ',');
			}

			skip_injail_passwd_check = iniparser_get_int_at_position(parser, section, "skip_injail_passwd_check", pos);

			free(section);
			
			if (allowed_jails == NULL) {
				syslog(LOG_ERR,"abort, no relevant section for user %s (%d) or group %s (%d) found in "CONFIGFILE, pw->pw_name,getuid(),gr->gr_name,getgid());
				exit(1);
			}
		} else {
			DEBUG_MSG("no relevant section found in configfile\n");
		}
		iniparser_close(parser);
	} else {
		DEBUG_MSG("no configfile "CONFIGFILE" ??\n");
		syslog(LOG_ERR,"abort, no config file "CONFIGFILE);
		exit(1);
	}
	
#ifdef OPEN_MAX
    i = OPEN_MAX;
#elif defined(NOFILE)
    i = NOFILE;
#else
    i = getdtablesize();
#endif
	while (--i > 2) {
		while (close(i) != 0 && errno == EINTR);
	}
	/* now make sure file descriptors 0 1 and 2 are valid before we (or a child) starts writing to it */
	while (1) {
		int fd;
		fd = open("/dev/null", O_RDWR);
		if (fd < 0)
			exit(10);
		if (fd > 2) {
			close(fd);
			break;
		}
	}
	
	
	/* check if the requested jail is allowed */
	{
		unsigned int allowed = 0;
		/* 'jail' has an ending slash */
		for (i=0;allowed_jails[i]!=NULL&&!allowed;i++) {
			allowed = dirs_equal(jail,allowed_jails[i]);
			DEBUG_MSG("allowed=%d after testing '%s' with '%s'\n",allowed,jail,allowed_jails[i]);
		}
		if (allowed!=1) {
			syslog(LOG_ERR,"abort, user %s (%d) is not allowed in jail %s",pw->pw_name, getuid(),jail);
			exit(21);
		}
	}
	/* test the jail */
	if (!basicjailissafe(jail)) {
		syslog(LOG_ERR, "abort, jail %s is not safe, check ownership and permissions for the jail inclusing system directories such as /etc, /lib, /usr, /dev, /sbin, and /bin", jail);
		exit(53);
	}
	
	if (chdir(jail) != 0) {
		syslog(LOG_ERR, "abort, chdir(%s) failed: %s",jail,strerror(errno));
		exit(19);
	} else {
		char test[1024];
		/* test if it really succeeded */
		if (getcwd(test, 1024)==NULL || !dirs_equal(jail, test)) {
			syslog(LOG_ERR, "abort, the current dir is %s after chdir(%s), but it should be %s",test,jail,jail);
			exit(21);
		}
	}
	
	syslog(LOG_INFO, "now entering jail %s for user %s (%d)", jail, pw->pw_name, getuid());
	
	/* do the chroot() call */
	if (chroot(jail)) {
		syslog(LOG_ERR, "abort, chroot(%s) failed: %s", jail, strerror(errno));
		exit(33);
	}
	
	if (use_capabilities) {
#ifdef HAVE_CAP_GET_PROC
		cap_t caps;
		cap_value_t capv[1];
		/* drop chroot capability, should we drop all other capabilities that may be used to escape from the jail too ?  */
		if ((caps = cap_get_proc()) == NULL) {
			syslog(LOG_ERR, "abort, failed to retrieve current capabilities: %s", strerror(errno));
			exit(101);
		}
		capv[0] = CAP_SYS_CHROOT;
		/* other capabilities that should/could be dropped:
		CAP_SETPCAP, CAP_SYS_MODULE, CAP_SYS_RAWIO, CAP_SYS_PTRACE, CAP_SYS_ADMIN */
		if (cap_set_flag(caps, CAP_PERMITTED, 1, capv, CAP_CLEAR)) {
			syslog(LOG_ERR, "abort, failed to set PERMITTED capabilities: %s", strerror(errno));
			exit(102);
		}
		if (cap_set_flag(caps, CAP_EFFECTIVE, 1, capv, CAP_CLEAR)) {
			syslog(LOG_ERR, "abort, failed to set effective capabilities: %s", strerror(errno));
			exit(103);
		}
		if (cap_set_flag(caps, CAP_INHERITABLE, 1, capv, CAP_CLEAR)) {
			syslog(LOG_ERR, "abort, failed to set INHERITABLE capabilities: %s", strerror(errno));
			exit(104);
		}
		if (cap_set_proc(caps)) {
			syslog(LOG_ERR, "abort, failed to apply new capabilities: %s", strerror(errno));
			exit(105);
		}
#else
		/* we should never get here */
		exit(333);
#endif		
	} else {
		/* drop all privileges, we first have to setgid(), 
			then we call setuid() */
		if (setgid(getgid())) {
			syslog(LOG_ERR, "abort, failed to set effective group ID %d: %s", getgid(), strerror(errno));
			exit(34);
		}
		if (setuid(getuid())) {
			syslog(LOG_ERR, "abort, failed to set effective user ID %d: %s", getuid(), strerror(errno));
			exit(36);
		}
	}

	if (!skip_injail_passwd_check){
		char *oldpw_name,*oldgr_name;
		oldpw_name = strdup(pw->pw_name);
		oldgr_name = strdup(gr->gr_name);
		
		if (user) {
			pw = getpwnam(user);
		} else {
			pw = getpwuid(getuid());
		}
		if (!pw) {
			syslog(LOG_ERR, "abort, failed to get user information in the jail for user ID %d: %s, check %s/etc/passwd",getuid(),strerror(errno),jail);
			exit(35);
		}
		DEBUG_MSG("got %s as pw_dir\n",pw->pw_dir);
		if (pw->pw_uid != getuid()) {
			syslog(LOG_ERR, "abort, got user information in the jail for user ID %d instead of user ID %d, check %s/etc/passwd",pw->pw_uid,getuid(),jail);
			exit(35);
		}
		gr = getgrgid(getgid());
		if (!gr) {
			syslog(LOG_ERR, "abort, failed to get group information in the jail for group ID %d: %s, check %s/etc/group",getgid(),strerror(errno),jail);
			exit(35);
		}
		if (strcmp(pw->pw_name, oldpw_name)!=0) {
			syslog(LOG_ERR, "abort, username %s differs from jail username %s for user ID %d, check /etc/passwd and %s/etc/passwd", oldpw_name, pw->pw_name, getuid(), jail);
			exit(37);
		}
		if (strcmp(gr->gr_name, oldgr_name)!=0) {
			syslog(LOG_ERR, "abort, groupname %s differs from jail groupname %s for group ID %d, check /etc/passwd and %s/etc/passwd", oldgr_name, gr->gr_name, getgid(), jail);
			exit(37);
		}
		free(oldpw_name);
		free(oldgr_name);
	}

	/* now execute the jailed shell */
	execv(executable, newargv);
	/* normally we wouldn't come to this bit of code */
	syslog(LOG_ERR, "ERROR: failed to execute %s for user %s (%d), check the permissions and libraries of %s%s",executable,pw->pw_name,getuid(),jail,executable);

	free(jail);
	exit(111);
}
